File: hot100/05_Linked_Lists/0146_LRU_Cache/solution.go
package main

import "fmt"

const (
	grn = "\033[32m"
	red = "\033[31m"
	bld = "\033[1m"
	rst = "\033[0m"
)

// ─── Solution ────────────────────────────────────────────────────────────────

type LRUCache struct {
	// TODO: add fields
}

func Constructor(capacity int) LRUCache {
	// TODO: implement
	return LRUCache{}
}

func (c *LRUCache) Get(key int) int {
	// TODO: implement
	return -1
}

func (c *LRUCache) Put(key, value int) {
	// TODO: implement
}

// ─── Test Harness ─────────────────────────────────────────────────────────────

func main() {
	passed, total := 0, 0

	check := func(got, want int, label string) {
		total++
		if got == want {
			fmt.Printf(grn+"✓"+rst+"  %-40s →  %d\n", label, got)
			passed++
		} else {
			fmt.Printf(red+"✗"+rst+"  "+bld+"%-40s"+rst+"    want=%d  got=%d\n", label, want, got)
		}
	}

	// Example 1 (from problem statement)
	{
		c := Constructor(2)
		c.Put(1, 1)
		c.Put(2, 2)
		check(c.Get(1), 1, "ex1: get(1)")
		c.Put(3, 3)
		check(c.Get(2), -1, "ex1: get(2) after eviction")
		c.Put(4, 4)
		check(c.Get(1), -1, "ex1: get(1) after eviction")
		check(c.Get(3), 3, "ex1: get(3)")
		check(c.Get(4), 4, "ex1: get(4)")
	}

	// Edge: capacity 1
	{
		c := Constructor(1)
		c.Put(1, 10)
		check(c.Get(1), 10, "cap1: get(1)=10")
		c.Put(2, 20)
		check(c.Get(1), -1, "cap1: get(1) evicted")
		check(c.Get(2), 20, "cap1: get(2)=20")
	}

	// Edge: update existing key
	{
		c := Constructor(2)
		c.Put(1, 1)
		c.Put(2, 2)
		c.Put(1, 10)
		c.Put(3, 3)
		check(c.Get(2), -1, "update: key 2 evicted")
		check(c.Get(1), 10, "update: key 1 = 10")
		check(c.Get(3), 3, "update: key 3 = 3")
	}

	// Edge: get promotes to MRU
	{
		c := Constructor(2)
		c.Put(1, 1)
		c.Put(2, 2)
		c.Get(1)
		c.Put(3, 3)
		check(c.Get(1), 1, "promote: key 1 survives")
		check(c.Get(2), -1, "promote: key 2 evicted")
	}

	fmt.Printf("\n%s\n", "--------------------------------------------")
	if passed == total {
		fmt.Printf(grn+"%d / %d passed\n"+rst, passed, total)
	} else {
		fmt.Printf(red+"%d / %d passed\n"+rst, passed, total)
	}
}

File: hot100/05_Linked_Lists/0146_LRU_Cache/solution.cpp
#include <bits/stdc++.h>
using namespace std;

#define GRN "\033[32m"
#define RED "\033[31m"
#define BLD "\033[1m"
#define RST "\033[0m"

// ─── Solution ────────────────────────────────────────────────────────────────

class LRUCache {
public:
    LRUCache(int capacity) {
        // TODO: implement
    }

    int get(int key) {
        // TODO: implement
        return -1;
    }

    void put(int key, int value) {
        // TODO: implement
    }
};

// ─── Test Harness ─────────────────────────────────────────────────────────────

int main() {
    int passed = 0, total = 0;

    auto check = [&](int got, int want, const string& label) {
        total++;
        if (got == want) {
            printf(GRN "✓" RST "  %-40s →  %d\n", label.c_str(), got);
            passed++;
        } else {
            printf(RED "✗" RST "  " BLD "%-40s" RST "    want=%d  got=%d\n",
                   label.c_str(), want, got);
        }
    };

    // Example 1 (from problem statement)
    {
        LRUCache c(2);
        c.put(1, 1);
        c.put(2, 2);
        check(c.get(1),    1, "ex1: get(1)");
        c.put(3, 3);
        check(c.get(2),   -1, "ex1: get(2) after eviction");
        c.put(4, 4);
        check(c.get(1),   -1, "ex1: get(1) after eviction");
        check(c.get(3),    3, "ex1: get(3)");
        check(c.get(4),    4, "ex1: get(4)");
    }

    // Edge: capacity 1
    {
        LRUCache c(1);
        c.put(1, 10);
        check(c.get(1),   10, "cap1: get(1)=10");
        c.put(2, 20);
        check(c.get(1),   -1, "cap1: get(1) evicted");
        check(c.get(2),   20, "cap1: get(2)=20");
    }

    // Edge: update existing key
    {
        LRUCache c(2);
        c.put(1, 1);
        c.put(2, 2);
        c.put(1, 10);
        c.put(3, 3);
        check(c.get(2),   -1, "update: key 2 evicted");
        check(c.get(1),   10, "update: key 1 = 10");
        check(c.get(3),    3, "update: key 3 = 3");
    }

    // Edge: get promotes to MRU
    {
        LRUCache c(2);
        c.put(1, 1);
        c.put(2, 2);
        c.get(1);
        c.put(3, 3);
        check(c.get(1),    1, "promote: key 1 survives");
        check(c.get(2),   -1, "promote: key 2 evicted");
    }

    printf("\n%s\n", string(44, '-').c_str());
    if (passed == total)
        printf(GRN "%d / %d passed\n" RST, passed, total);
    else
        printf(RED "%d / %d passed\n" RST, passed, total);
    return (passed == total) ? 0 : 1;
}
